### ISSUE TITLE
Empty tensor passed to optimizer prevents parameters update on CUDA device

### ISSUE THREAD

**Issue by thomasw21** [pytorch/pytorch#100701]
### 🐛 Describe the bug

Somehow passing an empty tensor in the optimizer prevents the update of normally updatable parameters. This happens only on CUDA device.

```python
from typing import Dict

import torch
from torch import nn
from torch.optim.adamw import AdamW

def assert_state_dict_equal(first: Dict[str, torch.Tensor], second: Dict[str, torch.Tensor]):
    for name, first_tensor in first.items():
        second_tensor = second[name]
        torch.testing.assert_close(first_tensor, second_tensor, atol=0, rtol=0)

def test_empty_tensor_in_optimizer(device: torch.device):
    model = nn.Linear(2, 3, bias=True, device=device)
    reference_model = nn.Linear(2, 3, bias=True, device=device)

    # sync parameters
    with torch.no_grad():
        model.weight.copy_(reference_model.weight)
        model.bias.copy_(reference_model.bias)

    empty_tensor = torch.tensor([], requires_grad=True, device=device)

    optimizer = AdamW(
        [
            model.weight,
            model.bias,
            empty_tensor,
        ]
    )
    reference_optimizer = AdamW(reference_model.parameters())

    random_input = torch.randn(5, 2, device=device)

    # Check that state dict are equal before optimizer step
    assert_state_dict_equal(model.state_dict(), reference_model.state_dict())

    model(random_input).sum().backward()
    empty_tensor.grad = torch.tensor([], requires_grad=False, device=device) # important
    reference_model(random_input).sum().backward()

    optimizer.step()
    reference_optimizer.step()

    # Check that state dict are equal after optimizer step
    assert_state_dict_equal(model.state_dict(), reference_model.state_dict())


def main():
    for device in [
        torch.device("cpu"),
        torch.device("cuda")
    ]:
        print(f"Testing device: {device}")
        test_empty_tensor_in_optimizer(device)
        print("Pass")


if __name__ == "__main__":
    main()
```

I would be expecting that adding empty parameters does not change the behaviour of other parameters passed down to the optimizer.

### Versions

PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1023-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 510.73.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Stepping:                        7
CPU MHz:                         3000.000
BogoMIPS:                        6000.00
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        48 MiB
L3 cache:                        71.5 MiB
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     KVM: Mitigation: VMX unsupported
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[pip3] triton==2.0.0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @janeyx99

**Comment by ngimel** [comment]
cc @janeyx99, @crcpar, with `foreach=False` test passes.

**Comment by janeyx99** [comment]
Sus..will take a look today

**Comment by janeyx99** [comment]
There was still a remaining edge case for this issue that recently popped up again. I will very soon open a PR resolving it, but the repro is:
```
import torch
shape_list = [torch.Size([1]), torch.Size([1]), torch.Size([0]), torch.Size([1]), torch.Size([0])]

params = [torch.tensor(torch.randn(s, device="cuda")) for s in shape_list]
params_clone = [p.clone().detach() for p in params]
updates = [torch.tensor(torch.ones(s, device="cuda")) for s in shape_list]

for i, param in enumerate(params_clone):
    param += updates[i]

torch._foreach_add_(params, updates)

index = 0
for p, p_clone in zip(params, params_clone):
    if not torch.allclose(p, p_clone):
        print(f"index {index} not equal")
    index += 1
print("total params: ", index)
```
Here, index 3 will not be correctly updated.

**Comment by janeyx99** [comment]
This should be fixed finally with the closing of #109402
